cps/elf.c

```
@@ -0,0 +1,259 @@
+	END(hdr->e_shstrndx, conv, &e_shstrndx_ntv);
+
+	/* Check whether there's enough room for the section headers. */
+	if (len < e_shoff_ntv + e_shnum_ntv * sizeof(sechdrs[0]))
+		return NULL;
+
+	sechdrs = (const ElfPERBIT(Shdr) *)
+		((const uint8_t *)hdr + e_shoff_ntv);
+
+	/* Check whether there's enough room for section with section names. */
+	END(sechdrs[e_shstrndx_ntv].sh_offset, conv, &sh_offset_ntv);
+	if (len < sh_offset_ntv)
+		return NULL;
+
+	/* Find symbol table. */
+	secstrings = (const char *)((const uint8_t *)hdr + sh_offset_ntv);
+
+	for (i = 1; i < e_shnum_ntv; i++) {
+		ElfPERBIT(Word) sh_name_ntv;
+		const char *found_secname;
+
+		END(sechdrs[i].sh_name, conv, &sh_name_ntv);
+		found_secname = secstrings + sh_name_ntv;
+
+		if (strcmp(secname, found_secname) == 0) {
+			ElfPERBIT(Word) sh_size_ntv;
+			END(sechdrs[i].sh_size, conv, &sh_size_ntv);
+			END(sechdrs[i].sh_offset, conv, &sh_offset_ntv);
+
+			/* Not enough room in file for this section. */
+			if (len < sh_offset_ntv + sh_size_ntv)
+				return NULL;
+
+			*secsize = sh_size_ntv;
+			return (const uint8_t *)hdr + sh_offset_ntv;
+		}
+	}
+
+	return NULL;
+}
+
+/* Find the next string in an ELF section. */
+static const char *
+next_string(const char *string, unsigned long *secsize)
+{
+	if (*secsize == 0)
+		return NULL;
+
+	/* Skip non-zero chars. */
+	while (string[0]) {
+		string++;
+		if ((*secsize)-- <= 1)
+			return NULL;
+	}
+
+	/* Skip any zero padding. */
+	while (!string[0]) {
+		string++;
+		if ((*secsize)-- <= 1)
+			return NULL;
+	}
+
+	return string;
+}
+
+static int
+load_string(struct elf_file *mod, const char *attr, char *val, size_t val_len)
+{
+	unsigned long size;
+	size_t attr_len = strlen(attr);
+	const char *strings;
+
+	/* Get strings from modinfo section. */
+	strings = get_section(mod, ".modinfo", &size);
+	if (strings == NULL) {
+		CPS_LOG(ERR, "Unable to get modinfo section from kernel module\n");
+		return -1;
+	}
+	RTE_VERIFY(size > 0);
```

Actually, `size` come from a file and it could be zero. So let's return an error when is zero.